frontend/server/app.test.ts

```diff
@@ -0,0 +1,333 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import * as os from 'os';
+import * as fs from 'fs';
+import * as path from 'path';
+import { PassThrough } from 'stream';
+
+import fetch from 'node-fetch';
+import * as requests from 'supertest';
+import { Client as MinioClient } from 'minio';
+import { Storage as GCSStorage } from '@google-cloud/storage';
+
+import { UIServer } from './app';
+import { loadConfigs } from './configs';
+import * as minioHelper from './minio-helper';
+import { getTensorboardInstance } from './k8s-helper';
+
+jest.mock('minio');
+jest.mock('node-fetch');
+jest.mock('@google-cloud/storage');
+jest.mock('./minio-helper');
+jest.mock('./k8s-helper');
+
+describe('UIServer', () => {
+  let app: UIServer;
+  const indexHtmlPath = path.resolve(os.tmpdir(), 'index.html');
+  const argv = ['node', 'dist/server.js', os.tmpdir(), '3000'];
+  const buildDate = new Date().toISOString();
+  const commitHash = 'abcdefg';
+  const indexHtmlContent = `
+<html>
+<head>
+  <script>
+  window.KFP_FLAGS.DEPLOYMENT=null
+  </script>
+  <script id="kubeflow-client-placeholder"></script>
+</head>
+</html>`;
+  const expectedIndexHtml = `
+<html>
+<head>
+  <script>
+  window.KFP_FLAGS.DEPLOYMENT="KUBEFLOW"
+  </script>
+  <script id="kubeflow-client-placeholder" src="/dashboard_lib.bundle.js"></script>
+</head>
+</html>`;
+
+  beforeAll(() => {
+    fs.writeFileSync(path.resolve(__dirname, 'BUILD_DATE'), buildDate);
+    fs.writeFileSync(path.resolve(__dirname, 'COMMIT_HASH'), commitHash);
+    fs.writeFileSync(indexHtmlPath, indexHtmlContent);
+  });
+
+  afterAll(() => {
+    fs.unlinkSync(path.resolve(__dirname, 'BUILD_DATE'));
+    fs.unlinkSync(path.resolve(__dirname, 'COMMIT_HASH'));
+    fs.unlinkSync(indexHtmlPath);
+  });
+
+  beforeEach(() => {
+    (MinioClient as any).mockClear();
+    (fetch as any).mockClear();
+    (GCSStorage as any).mockClear();
+    (getTensorboardInstance as any).mockClear();
+  });
+
+  afterEach(() => {
+    app && app.close();
+  });
+
+  it('api server is not ready', done => {
```

Comment:
@eterna2 thanks for these great efforts in testing! They help a lot!

I have to add some style guide here. Recommended test message structure is the following:
* describe messages describe what subject or conditions you are testing.
* you can use describe nesting to divide tests by groups, e.g.
```
describe('UIServer', () => {
  describe('index.html', () => {
    it('responds', ...);
    it('responds with kubeflow flag and client injected when in kubeflow deployment', ...);
  });
});
```
* it messages describe expected behavior after certain actions. Usually it messages should read like an english sentence like `it('reports api server is not ready', () => {...})`.
* it message should convey roughly about expected behavior too. `it` refers to the subject you are testing.

The guideline is consistent with existing UI unit tests in this repo and google internal guides (unfortunately they are not public), but I think this is fairly common practice. I don't have a good resource on this, you can grasp an idea of that by reading this: https://kentcdodds.com/blog/write-fewer-longer-tests (read his test messages, your tests are structured at a good size).

Naming things is hard... which we all know. So if you don't feel comfortable changing this, I can follow up after this is merged too. What do you think?

Reply:
sure. I can make the changes.

Reply:
Thank you! That would be awesome!